### ISSUE TITLE
API: exposing CoordinateSequence in Shapely 2.0

### ISSUE THREAD

**Issue by jorisvandenbossche** [Toblerity/Shapely#984]
One of the questions I run into while doing the Geometry classes refactor (subclassing PyGEOS geometry -> https://github.com/Toblerity/Shapely/pull/983), is what to do with CoordinateSequence.

So right now, if you access the `.coords` of a Shapely Geometry you get a `CoordinateSequence` object (at least for Point, LineString and LinearRing, the geometry types that are directly backed by a coordinate sequence).

Some examples usages that are possible now:

```python
>>> from shapely.geometry import LineString  
>>> geom = LineString([(0, 0), (0, 1), (1, 1)])
>>> print(geom) 
LINESTRING (0 0, 0 1, 1 1)

>>> geom.coords   
<shapely.coords.CoordinateSequence at 0x7f8663bd5438>

# slicing gives a list of tuples (this is done often in the tests)
>>> geom.coords[:]   
[(0.0, 0.0), (0.0, 1.0), (1.0, 1.0)]

# accessing single element gives (x, y) or (x, y, z) tuple
>>> geom.coords[0] 
(0.0, 0.0)

# iterating over coordinate tuples
>>> for x, y in geom.coords: 
...     print(x, y) 
0.0 0.0
0.0 1.0
1.0 1.0

# the xy attribute gives two 1D arrays (from the stdlib array module)
>>> geom.coords.xy
(array('d', [0.0, 0.0, 1.0]), array('d', [0.0, 1.0, 1.0]))

# convert to a 2D numpy array
>>> np.asarray(geom.coords) 
array([[0., 0.],
       [0., 1.],
       [1., 1.]])
```

Apart from the exact API (eg tuples and not numpy arrays), the current implementation also ensures you can *lazily* iterate over the coordinates or get a specific coordinate pair (instead of needing to convert all coordiantes of the full Geometry to a list/array first).

For Shapely 2.0, I think there are some different options (from most to least compatible with current shapely):

1) We also implement a `CoordinateSequence` extension type that maps to `GEOSCoordSequence` (like `pygeos.Geometry` wraps a `GEOSGeometry`), so that we can keep the exact same interface and the lazy conversion aspect.  
  However, this also gives quite some additional code complexity (implementing yet another extension type in C, although the code itself will be quite similar as the Geometry extension type).

2) We keep the python `CoordinateSequence` class for the return value of the `.coords` property to mimic the current API, but we let it hold the full coordinates array instead of a GEOCoordSequence objet (so when accessing `geom.coords`, we get all the coordinates of the geometry as a numpy array, and put that in a CoordinateSequence wrapper class). This can preserve the API, but looses the lazy aspect.

3) We simply return the numpy array of coordinates (what you would get now with `np.asarray(geom.coords)`) instead of having the CoordinateSequence class. Behavioural consequence: we need to deprecate the `xy` attribute (but you can of course slice the 2D array to get 2 1D arrays), and accessing coordinates or iterating over the coords will now gives a numpy array instead of a tuple for coordinate pairs (eg `np.array([0., 0.])` instead of `(0.0, 0.0)`). 
  Potential option 3b) would be to convert the numpy array to a list of tuples, and return that from `.coords`. This preserves the behaviour of getting a tuple for coordinate pairs in getitem/iteration. But on the other hand, makes getting an array of coordianates of a large Geometry less efficient.

Personally, I am not sure how important the lazy aspect is (the conversion to a numpy array of coordinates will also be more efficient with pygeos). 
In principle, we could also implement a "get nth coordinate pair" function in C that works directly on a supported geometry instead of a CoordinateSequence, and that way we could keep a lazy iteration also in option 2, while still have the efficient conversion to a full numpy array of coordinates.

For now, in my refactor of the Geometry classes (https://github.com/Toblerity/Shapely/pull/983), I implemented option 2 as this was the easiest to do while getting existing tests passing.

cc @caspervdw @sgillies @mwtoews

**Comment by sgillies** [comment]
Thanks for the great discussion @jorisvandenbossche . I'd prefer to keep the coords attribute mostly as it is. I'm ambivalent about the implementation but wonder if it won't be harder to keep the geometry coordinates immutable if we implement them using a naturally mutable structure like a numpy array or list.

**Comment by jorisvandenbossche** [comment]
> wonder if it won't be harder to keep the geometry coordinates immutable if we implement them using a naturally mutable structure like a numpy array or list.

We don't really "implement" the coordinates, as it is only providing *access* to the underlying GEOS CoordSequence object. So those numpy arrays or lists are always a copy of the coordinates, and don't allow mutating the actual CoordSequence (the same is true for the current Shapely `CoordinateSequence`, which uses GEOSCoordSeq_getX_r et al to get to the coordinates one by one). 

Now, if we directly return a numpy array or list as `.coords` attribute, you can of course mutate that array/list without having any effect, which might be confusing. So that's maybe another argument for keeping a CoordinateSequence class wrapping the numpy array of coordinates, so we can disallow `__setitem__` on the `.coords` attribute.

**Comment by caspervdw** [comment]
Implementing an additional CoordSequence C extension type is not necessary. We can access the CoordSequence really efficiently with `GEOSGeom_getCoordSeq_r` whenever we need.

It seems to me that we should implement some methods that would support an array interface on geometries (which ones?). At least `get_coordinate(geometry, n) -> (x, y[, z])`, or maybe a more generic `get_coordinate_range(geometry, start, stop, step=1) -> ndarray` will make it possible to keep the shapely CoordinateSequence API as is.

The first one (scalar coordinate retrieval) is really not hard to make.

**Comment by jorisvandenbossche** [comment]
Indeed, with a `get_coordinate(geometry, n)` function, that should be sufficient to fully mimic the current API.

**Comment by sgillies** [comment]
What do you think about supporting slicing?

**Comment by jorisvandenbossche** [comment]
Slicing can be supported with a `get_coordinate(geometry, n)` as well, I think? Similarly as how slicing is now also implemented by looping and the single coordinate accessors:

https://github.com/Toblerity/Shapely/blob/73626bce8e99b3f251ad874a41d97d2384e734ce/shapely/coords.py#L89-L100
